### ISSUE TITLE
`E0308` false positive regression

### ISSUE THREAD

**Issue by LucaCappelletti94** [rust-lang/rust-analyzer#21429]
Basically, since a recent version (0.3.2735-standalone ) there was a regression in which Rust Analyzer started to raise `EO308` errors on all `borrow` types, raising false positive errors such as the following in [`sql-traits`](https://github.com/earth-metabolome-initiative/sql-traits). There are a few much older closed issues from 2023/2024 describing the same issue, so I believe this is a regression. These issues are of course not present in `cargo check`, in either stable or nightly.

```rust
expected &<&'db <<Self as TableLike>::DB as DatabaseLike>::CheckConstraint as CheckConstraintLike>::DB, found &'db <Self as TableLike>::DB
expected &<<&<<Self as ColumnLike>::DB as DatabaseLike>::CheckConstraint as CheckConstraintLike>::DB as DatabaseLike>::Column, found &<<Self as ColumnLike>::DB as DatabaseLike>::Column
```

**most recent tested versions that work**:

* 0.3.2711-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2711-linux-x64/server/rust-analyzer] - 1 month ago
* 0.3.2719-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2719-linux-x64/server/rust-analyzer] - 3 weeks ago
* 0.3.2727-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2727-linux-x64/server/rust-analyzer] - 2 weeks ago

**broken rust-analyzer versions**:

* 0.3.2735-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2735-linux-x64/server/rust-analyzer] - 1 week ago
* 0.3.2743-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2743-linux-x64/server/rust-analyzer] - 4 days ago
* 0.4.2747-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.4.2747-linux-x64/server/rust-analyzer] - pre-release version

**rustc version**: rustc 1.92.0 (ded5c06cf 2025-12-08)

**editor or extension**: VSCode

**relevant settings**: unsure whether there are any relevant settings in this context

**repository link (if public, optional)**: the repository present this issue is [sql-traits](https://github.com/earth-metabolome-initiative/sql-traits)

**Comment by ChayimFriedman2** [comment]
Do you have experimental diagnostics enabled? If not, that's not a regression, we just enabled the diagnostic by default.

**Comment by ChayimFriedman2** [comment]
Where exactly are the errors?

**Comment by LucaCappelletti94** [comment]
> Do you have experimental diagnostics enabled? If not, that's not a regression, we just enabled the diagnostic by default.

I am not sure what that is. How can I verify whether it is enabled or disabled? Are you referring to these? BTW, I am currently running `0.3.2727-standalone` as it was the most recent version which did not raise false positives.

<img width="1156" height="210" alt="Image" src="https://github.com/user-attachments/assets/f590e434-5d7c-470b-a15f-34ff18ad13d2" />

**Comment by ChayimFriedman2** [comment]
Yes, they are disabled for you.

<s>Does the error also reproduce in prerelease?</s> Just saw you said not.

**Comment by LucaCappelletti94** [comment]
> Where exactly are the errors?

Basically, in version `0.3.2735-standalone` and later, Rust Analyzer raises `E0308` errors in several cases where it fails to match two associated types, while `cargo check` succeeds. **The failure is also present in the pre-released version 0.4.2747-standalone.**

Here is an example of a complete error as produced in version 0.3.2743-standalone [~/.vscode-server/extensions/rust-lang.rust-analyzer-0.3.2743-linux-x64/server/rust-analyzer], raise on [this code here](https://github.com/earth-metabolome-initiative/sql-traits/blob/28363f9369607f703c3973c4c51e336986b623d9/src/traits/column.rs#L352-L362):

```rust
expected &<&<<Self as ColumnLike>::DB as DatabaseLike>::ForeignKey as ForeignKeyLike>::DB, found &'db <Self as ColumnLike>::DBrust-analyzer[E0308](https://doc.rust-lang.org/stable/error_codes/E0308.html)
database: &'db <Self as ColumnLike>::DB
```

On code:

```rust
fn foreign_keys<'db>(
        &'db self,
        database: &'db Self::DB,
    ) -> impl Iterator<Item = &'db <Self::DB as DatabaseLike>::ForeignKey>
    where
        Self: 'db,
    {
        ColumnLike::table(self, database).foreign_keys(database).filter(move |fk| {
            fk.host_columns(database).map(Borrow::borrow).any(|col: &Self| col == self)
        })
    }
```

Where specifically it fails to resolve that `database` is, [by trait bound in `DatabaseLike`](https://github.com/earth-metabolome-initiative/sql-traits/blob/28363f9369607f703c3973c4c51e336986b623d9/src/traits/database.rs#L16-L28), the same type.

<img width="1259" height="251" alt="Image" src="https://github.com/user-attachments/assets/f1939bf4-3a35-4f97-ba5b-0519a741a037" />

**Comment by LucaCappelletti94** [comment]
> Does the error also reproduce in prerelease?

Yes, it also does reproduce in the pre-release version `0.4.2747-standalone`

**Comment by LucaCappelletti94** [comment]
I checked whether the aforementioned experimental diagnostics were enabled by default in the pre-release version `0.4.2747-standalone`. The diagnostic seem to be disabled, yet the false positive persists. I believe, therefore, that the false positive is not caused by these experimental features but something else and is therefore a regression.

<img width="824" height="112" alt="Image" src="https://github.com/user-attachments/assets/5c862b03-83df-41df-bc9e-5842ea31dfc4" />

**Comment by ChayimFriedman2** [comment]
The experimental diagnostics config was not enabled by default, by the diagnostic stopped being experimental.

**Comment by ChayimFriedman2** [comment]
Minimal reproduction:
```rust
use std::borrow::Borrow;

pub trait DatabaseLike {
    type Column: ColumnLike<DB = Self>;
    type ForeignKey: ForeignKeyLike<DB = Self>;
}

pub trait ForeignKeyLike:
    Borrow<<<Self as ForeignKeyLike>::DB as DatabaseLike>::ForeignKey>
{
    type DB: DatabaseLike<ForeignKey: Borrow<Self>>;

    fn host_columns<'db>(&'db self, database: &'db Self::DB)
    where
        Self: 'db;
}

pub trait ColumnLike: Borrow<<<Self as ColumnLike>::DB as DatabaseLike>::Column> {
    /// The type of the database that this column belongs to.
    type DB: DatabaseLike<Column: Borrow<Self>>;

    fn foo(&self) -> &&<<Self as ColumnLike>::DB as DatabaseLike>::ForeignKey {
        todo!()
    }

    fn foreign_keys<'db>(&'db self, database: &'db Self::DB)
    where
        Self: 'db,
    {
        let fk = self.foo();
        fk.host_columns(database);
    }
}

impl<'a, C> ColumnLike for &'a C
where
    C: ColumnLike,
    Self: Borrow<<<C as ColumnLike>::DB as DatabaseLike>::Column>,
    <<C as ColumnLike>::DB as DatabaseLike>::Column: Borrow<&'a C>,
{
    type DB = C::DB;
}
```

**Comment by ChayimFriedman2** [comment]
More minimal reproduction:
```rust
trait DatabaseLike {
    type ForeignKey: ForeignKeyLike<DB = Self>;
}

trait ForeignKeyLike {
    type DB: DatabaseLike;

    fn host_columns(&self, database: &Self::DB);
}

trait ColumnLike {
    type DB: DatabaseLike;

    fn foo(&self) -> &&<<Self as ColumnLike>::DB as DatabaseLike>::ForeignKey {
        todo!()
    }

    fn foreign_keys(&self, database: &Self::DB) {
        let fk = self.foo();
        fk.host_columns(database);
    }
}
```

**Comment by anson-vandoren** [comment]
Not sure it matters now that there's a PR already, but this actually starts on v0.3.2727 for me, but does not have the regression in v3.3.2719

**Comment by LucaCappelletti94** [comment]
@anson-vandoren if you can, please do provide a minimal reproduction so to check that the two false positives are indeed the same one. I don't get any E0308 error in `v0.3.2727`, so it might be a similar but different bug.

**Comment by jayprajapati-inventyv** [comment]
<img width="793" height="241" alt="Image" src="https://github.com/user-attachments/assets/ccc7edc3-855a-414c-9029-8a3a9837e092" />

I am getting error code E0308 from rust-analyzer in VS Code, but cargo check does not report any errors.

The project builds successfully with Cargo, but rust-analyzer shows a type mismatch error. I have already checked that proc macros are enabled and verified my Rust toolchain configuration.

Could you please help me identify why rust-analyzer is reporting this error and how to resolve it?
